### PR TITLE
CLM-19 | add relationship migration

### PR DESCRIPTION
**Pull request by peterlu0ng**
Add relationship migration

- References user table
- Makes a friend reference on the table
- Enum status with the two statuses: "Friends" & "Blocked"

### CODE REVIEW

File: db/migrate/20240817024535_create_relationships.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateRelationships < ActiveRecord::Migration[7.1]
+  def change
+    create_enum :friendship_status, ['Friends', 'Blocked']
+    create_table :relationships do |t|
+      t.references :user, null: false, foreign_key: true
+      t.references :friend, null: false, foreign_key: { to_table: :users }
+      t.enum :friendship, enum_type: 'friendship_status', null: false
+      t.timestamps
```
Comment (builtbyproxy) [inline db/migrate/20240817024535_create_relationships.rb:8]: Why do we want timestamps?

**Review by builtbyproxy — CHANGES_REQUESTED** [review]
As discussed on our call, I would do this a little differently; 

1. Define the `status` as an integer and then the enum itself in the model: 
```
enum status: { pending: 0, accepted: 1, declined: 2, blocked: 3 }
```
You might not want declined for example, you might want to just delete the relationship

2. I've asked ChatGPT and agree with it's thoughts: 

Migration: 
```ruby
class CreateRelationships < ActiveRecord::Migration[7.0]
  def change
    create_table :relationships, id: :uuid do |t|
      t.uuid :user_id_a, null: false, foreign_key: { to_table: :users }
      t.uuid :user_id_b, null: false, foreign_key: { to_table: :users }
      t.integer :status, null: false

      t.timestamps
    end

    # Ensure that the combination of user_id_a and user_id_b is unique
    add_index :relationships, [:user_id_a, :user_id_b], unique: true

    # Optionally, you can enforce user_id_a < user_id_b to avoid duplication
    add_check_constraint :relationships, 'user_id_a < user_id_b', name: 'user_id_order_check'
  end
end
```

Model: 
```ruby
class Relationship < ApplicationRecord
  enum status: { pending: 0, accepted: 1, declined: 2, blocked: 3 }

  belongs_to :user_a, class_name: 'User', foreign_key: 'user_id_a'
  belongs_to :user_b, class_name: 'User', foreign_key: 'user_id_b'

  validates :user_a, presence: true
  validates :user_b, presence: true
  validates :status, presence: true

  validate :not_self_relationship
  validate :unique_relationship

  private

  def not_self_relationship
    errors.add(:user_b, "can't be the same as user_a") if user_id_a == user_id_b
  end

  def unique_relationship
    if Relationship.where(user_id_a: user_id_a, user_id_b: user_id_b).or(Relationship.where(user_id_a: user_id_b, user_id_b: user_id_a)).exists?
      errors.add(:base, "Relationship already exists")
    end
  end
end
```

I don't agree entirely with this. I think the `add_check_constraint` could be a bit overkill. But take this as a starting point.

File: db/migrate/20240817024535_create_relationships.rb (hunk 1 above)
Comment (peterlu0ng) [reply db/migrate/20240817024535_create_relationships.rb]: I've removed timestamps, doesn't really make sense for to know when they are friends

**Review by builtbyproxy — APPROVED** [review]
As a tiny nit; You've merged #8 to this PR instead of `main` which has muddied the waters of this PR. It's not a massive issue here because of the size of the context free but in future the 'better' path is #8  -> `main` -> This PR.

**Pull request merged**